event/subscriber.go

```diff
@@ -0,0 +1,89 @@
+package event
+
+import (
+	"context"
+	"fmt"
+	"github.com/medibloc/panacea-doracle/config"
+	log "github.com/sirupsen/logrus"
+	rpchttp "github.com/tendermint/tendermint/rpc/client/http"
+	"time"
+)
+
+type PanaceaSubscriber struct {
+	Subscriber string
+	Client     *rpchttp.HTTP
+}
+
+// NewSubscriber generates a rpc http client with websocket address.
+func NewSubscriber(conf *config.Config) (*PanaceaSubscriber, error) {
+	client, err := rpchttp.New(conf.Panacea.WSAddr, "/websocket")
+	if err != nil {
+		return nil, err
+	}
+
+	err = client.Start()
+	if err != nil {
+		return nil, err
+	}
+
+	return &PanaceaSubscriber{
+		Subscriber: conf.BaseConfig.Subscriber,
+		Client:     client,
+	}, nil
+}
+
+type PanaceaEventStatus int32
+
+const (
+	RegisterOracle PanaceaEventStatus = 1
+)
+
+func (s *PanaceaSubscriber) Run(event ...PanaceaEventStatus) error {
+	log.Infof("start panacea event subscriber")
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+
+	for _, e := range event {
+		convertedEvent := convertEventStatusToEvent(e)
+		query := convertedEvent.GetEventType() + "." + convertedEvent.GetEventAttributeKey() + "=" + convertedEvent.GetEventAttributeValue()
+		txs, err := s.Client.Subscribe(ctx, s.Subscriber, query)
+		if err != nil {
+			return err
+		}
```

Comment:
It's better to create a context (with timeout) right before calling `s.Client.Subscribe`, if you intended to set the 1s timeout for sending a ''subscribe' request to the Panacea.
If you create a context outside of the loop and reuse it for 'subscribe' calls, the timeout error will occur unless all 'subscribe' calls are done in 1s.

Reply:
When I changed like following code, there has a warning in `defer.cancel()`.
The warning said `Possible resource leak, 'defer' is called in the 'for' loop`. Is it okay to change like following code?
```go
for _, e := range event {
		convertedEvent := convertEventStatusToEvent(e)
		query := convertedEvent.GetEventType() + "." + convertedEvent.GetEventAttributeKey() + "=" + convertedEvent.GetEventAttributeValue()

		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
		defer cancel()

		txs, err := s.Client.Subscribe(ctx, "", query)
		if err != nil {
			return err
		}
```

Reply:
Yeah. The warning matters. Let me check how the `s.Client.Subscribe` is implemented. 1 sec.

Reply:
Alright. I just checked how the internal of `Subscribe(ctx, ...)` works. The `ctx` is used only when sending a packet which initializes the subscription. After the subscription is established, the `ctx` is not used. It means that the subscriber will subscribe new events endlessly as we intend. All good.

Nevertheless, it's important to set a timeout using `ctx` as you did, so that we can get a timeout error if it takes a long time to send a initialization packet.
To avoid the warning in terms of `defer`, I would recommend you to define a new function like:
```go
func subsribe(query string) {
          ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
          defer cancel()

          txs, err := s.Client.Subscribe(ctx, "", query)
          if err != nil {
              return err
          }

          go func(e Event) {
              for t := range txs {
                  fmt.Println("got ", t.Events)
                  _ = e.EventHandler(t)
              }
          }(e)
}
```

Then, you can call this function in your for loop.

Reply:
Actually, you don't need to set timeouts for now when communicating with Tendermint. There are lots of logics in our doracle which communicate with Tendermint, but we aren't setting any timeout for now. Later, we have to define a new parameter `panacea_write_timeout` in our `config.toml`, and use it everywhere. But, let's go with your code for now.

Reply:
Okay I got it!. Thank you.
I refactored in [67719da](https://github.com/medibloc/panacea-doracle/pull/15/commits/67719daf9bf6e90041cceae9d5adeb4c60315bc8).